consensus/src/simplex/actors/voter/actor.rs

```diff
@@ -161,55 +182,134 @@ impl<E: Clock, S: Scheme, D: Digest> Round<E, S, D> {
+            }
         }
+
+        // Certificate proposals are considered verified (they have 2f+1 agreement)
+        if matches!(
+            self.proposal_status,
+            ProposalStatus::None | ProposalStatus::Unverified
+        ) {
+            debug!(?proposal, "setting verified proposal from certificate");
+            self.proposal_status = ProposalStatus::Verified;
+        }
+
+        self.proposal = Some(proposal);
+        None
     }
 
-    async fn add_verified_notarize(&mut self, notarize: Notarize<S, D>) {
-        if self.proposal.is_none() {
+    async fn add_verified_notarize(&mut self, notarize: Notarize<S, D>) -> Option<S::PublicKey> {
+        // Check for equivocation
+        if self.proposal_status == ProposalStatus::Replaced {
+            return None;
+        }
+
+        // Check for conflicting proposal
+        if let Some(proposal) = self.proposal.as_ref() {
+            if proposal != &notarize.proposal {
+                // Mark status as replaced and extract equivocator (if known)
+                let equivocator = self.leader.as_ref().map(|leader| leader.key.clone());
+                warn!(
+                    ?equivocator,
+                    ?proposal,
+                    ?notarize.proposal,
+                    "notarize conflicts with certificate proposal (equivocation detected)"
+                );
+                self.proposal_status = ProposalStatus::Replaced;
+                self.notarizes.clear();
+                self.finalizes.clear();
+                return equivocator;
+            }
+        } else {
             self.proposal = Some(notarize.proposal.clone());
+            self.proposal_status = ProposalStatus::Unverified;
         }
         self.notarizes.insert(notarize);
+        None
     }
 
     async fn add_verified_nullify(&mut self, nullify: Nullify<S>) {
         // We don't consider a nullify vote as being active.
         self.nullifies.insert(nullify);
     }
 
-    async fn add_verified_finalize(&mut self, finalize: Finalize<S, D>) {
-        if self.proposal.is_none() {
+    async fn add_verified_finalize(&mut self, finalize: Finalize<S, D>) -> Option<S::PublicKey> {
+        // Check for equivocation
+        if self.proposal_status == ProposalStatus::Replaced {
+            return None;
+        }
+
+        // Check for conflicting proposal
+        if let Some(proposal) = &self.proposal {
+            if proposal != &finalize.proposal {
+                // Mark status as replaced and extract equivocator (if known)
+                let equivocator = self.leader.as_ref().map(|leader| leader.key.clone());
+                warn!(
+                    ?equivocator,
+                    ?proposal,
+                    ?finalize.proposal,
+                    "finalize conflicts with certificate proposal (equivocation detected)"
+                );
+                self.proposal_status = ProposalStatus::Replaced;
+                self.notarizes.clear();
```

Comment:
This is required for the nuanced case where we get a cert, vote off that cert, get a message from batcher that the proposal from the leader is different.